stl/src/filesystem.cpp

```diff
@@ -353,45 +354,68 @@ void __stdcall __std_fs_directory_iterator_close(_In_ const __std_fs_dir_handle
             __std_win_error _Last_error = _Translate_CreateFile_last_error(_Source_handle._Get());
             if (_Last_error != __std_win_error::_Success) {
                 return {false, _Last_error};
             }
 
-            const _STD _Fs_file _Target_handle(__vcp_CreateFile(
-                _Target, FILE_READ_ATTRIBUTES | FILE_WRITE_DATA, 0, nullptr, OPEN_EXISTING, 0, nullptr));
+            const _STD _Fs_file _Target_handle(
+                __vcp_CreateFile(_Target, FILE_READ_ATTRIBUTES, _Share_mode, nullptr, OPEN_EXISTING, 0, nullptr));
             _Last_error = _Translate_CreateFile_last_error(_Target_handle._Get());
             if (_Last_error != __std_win_error::_Success) {
-                // Also handles the equivalent(from, to) error case
                 return {false, _Last_error};
             }
 
-            // If we get here, we did the equivalent(from, to) test. If we were only asked to skip existing, we're done
-            if (_Options == __std_fs_copy_options::_Skip_existing) {
-                return {false, __std_win_error::_Success};
-            }
+            bool _Do_copy = false;
+            if (_Options == __std_fs_copy_options::_Update_existing) {
+                long long _Source_last_write_time;
+                _Last_error = _Get_last_write_time_by_handle(_Source_handle._Get(), &_Source_last_write_time);
+                if (_Last_error != __std_win_error::_Success) {
+                    return {false, _Last_error};
+                }
 
-            // Test for update_existing
-            long long _Source_last_write_time;
-            _Last_error = _Get_last_write_time_by_handle(_Source_handle._Get(), &_Source_last_write_time);
-            if (_Last_error != __std_win_error::_Success) {
-                return {false, _Last_error};
-            }
+                long long _Target_last_write_time;
+                _Last_error = _Get_last_write_time_by_handle(_Target_handle._Get(), &_Target_last_write_time);
+                if (_Last_error != __std_win_error::_Success) {
+                    return {false, _Last_error};
+                }
 
-            long long _Target_last_write_time;
-            _Last_error = _Get_last_write_time_by_handle(_Target_handle._Get(), &_Target_last_write_time);
-            if (_Last_error != __std_win_error::_Success) {
-                return {false, _Last_error};
+                if (_Target_last_write_time < _Source_last_write_time) { // _Target is newer, so update_existing
+                    _Do_copy = true;
+                }
             }
 
-            if (_Source_last_write_time <= _Target_last_write_time) { // _Target is newer, so don't update_existing
+            if (!_Do_copy) {
+                // We only need to test `equivalent()` if we _aren't_ going to `CopyFileW()`,
+                // since that call will fail with an `ERROR_SHARING_VIOLATION` anyways.

```

Comment:
This is Desktop and Recent only so you need this fallback: https://github.com/microsoft/STL/blob/4cd7ff723c1a28ae2a757b4856bf6ca795492b10/stl/src/filesystem.cpp#L461

Reply:
I see - [`GetFileInformationByHandleEx`](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getfileinformationbyhandleex) is documented as:

> **Minimum supported client** Windows Vista [desktop apps | UWP apps]
> **Minimum supported server** Windows Server 2008 [desktop apps | UWP apps]

However, [`FILE_ID_INFO`](https://docs.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_info) says:

> **Minimum supported client** None supported
> **Minimum supported server** Windows Server 2012 [desktop apps only]

Excellent catch! It looks like the proper thing to do is to call the `__std_fs_get_file_id()` that you cited (instead of trying to replicate what it does).